### ISSUE TITLE
Compressed Fits Files Padding

### ISSUE THREAD

**Issue by cmccully** [astropy/astropy#9072]
This is partially trying to understand how astropy works with fpacked files So it can include the label of question. I think I mostly understand how to read the fits files, but it would be nice to have the convenience functions I have below somewhere in astropy. On the other hand, the file padding is quite baffling to me. Any clarity on this issue would be appreciated.

Throughout this we will use from https://archive.lco.global/?q=a&RLEVEL=91&PROPID=&INSTRUME=&OBJECT=&SITEID=lsc&TELID=1m0a&FILTER=&OBSTYPE=STANDARD&EXPTIME=&BLKUID=&REQNUM=&basename=lsc1m009-fa04-20190429-0061-s91&start=2019-04-01%2000%3A00&end=2019-04-30%2023%3A00&id=&public=true

I recommend downloading the data and funpacking the file to be used as the raw data. This way you don't need to worry about quantization noise differing between runs.

```python
import astropy
print(astropy.version.version)
from astropy.io import fits
print(fits.compression.CFITSIO_VERSION)
```

    3.2.1
    3.45

```python
import os
from glob import glob
import numpy as np
```

Reading compressed files:
======================

In general, it would be nice to have a convenience function that reads in an fpacked file into the same structure as if it were not compressed. Currently we have the following behavior:

```python
files = glob('*fits*')
print(files)
raw_filename = files[0]
```

    ['lsc1m009-fa04-20190429-0061-s91.fits']

```python
uncompressed_hdulist = fits.open(raw_filename)
uncompressed_hdulist.info()
```

    Filename: lsc1m009-fa04-20190429-0061-s91.fits
    No.    Name      Ver    Type      Cards   Dimensions   Format
      0  SCI           1 PrimaryHDU     273   (4096, 4096)   float32   
      1  CAT           1 BinTableHDU    157   306R x 37C   [D, D, D, D, K, K, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, K, D, D]   
      2  BPM           1 ImageHDU        10   (4096, 4096)   uint8   

Now we will compress the file using fpack, and it back in using astropy (we will compare compressing files below).

`1.7.0 (Dec 2013) CFITSIO version 3.450`

```python
os.system('/home/cmccully/src/cfitsio-3.45/fpack -r -q2048 64 lsc1m009-fa04-20190429-0061-s91.fits && mv lsc1m009-fa04-20190429-0061-s91.fits.fz fpacked.fits.fz')
```

    0

To get rid of differences due to quantization noise, we also funpack the fpacked file for comparision:

```python
os.system('funpack fpacked.fits.fz')
funpacked_hdulist = fits.open('fpacked.fits')
```

```python
fpacked_hdulist = fits.open('fpacked.fits.fz')
fpacked_hdulist.info()
```

    Filename: fpacked.fits.fz
    No.    Name      Ver    Type      Cards   Dimensions   Format
      0  PRIMARY       1 PrimaryHDU       8   ()      
      1  SCI           1 CompImageHDU    280   (4096, 4096)   float32   
      2  CAT           1 BinTableHDU    157   306R x 37C   [D, D, D, D, K, K, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, K, D, D]   
      3  BPM           1 CompImageHDU     10   (4096, 4096)   uint8   

Note that if I just try to write this back out to a file, we get unexpected behavior:

```python
fpacked_hdulist['BPM'].data = np.array(fpacked_hdulist['BPM'].data, np.uint8)
fpacked_hdulist.writeto('fpack_straight_out.fits.fz', overwrite=True)
```

```python
!du -hs *.fz
```

    19M	fpacked.fits.fz
    50M	fpack_straight_out.fits.fz

Besides needing to fix the cast to uint8 (see more below), it appears to be bigger than the original.
_______________________________________________________________________________________

That aside, the file structure is different so even the index of the extension is not preserved.

To convert between the two, you have to do something like the following:

```python
def unpack(compressed_hdulist: fits.HDUList) -> fits.HDUList:
    FITS_MANDATORY_KEYWORDS = ['SIMPLE', 'BITPIX', 'NAXIS', 'EXTEND', 'COMMENT', 'CHECKSUM', 'DATASUM']
    # If the primary fits header only has the mandatory keywords, then we throw away that extension 
    # and extension 1 gets moved to 0
    # Otherwise the primary HDU is kept
    move_1_to_0 = True
    for keyword in compressed_hdulist[0].header:
        if keyword not in FITS_MANDATORY_KEYWORDS:
            move_1_to_0 = False
            break
    if not move_1_to_0 or not isinstance(compressed_hdulist[1], fits.CompImageHDU):
        primary_hdu = fits.PrimaryHDU(data=None, header=compressed_hdulist[0].header)
    else:
        data_type = compressed_hdulist[1].data.dtype
        if 'int' == str(data_type)[:3]:
            data_type = getattr(np, 'u' + data_type)
            data = np.array(compressed_hdulist[1].data, getattr(np, data_type))
        else:
            data = compressed_hdulist[1].data
        primary_hdu = fits.PrimaryHDU(data=data, header=compressed_hdulist[1].header)
        primary_hdu.header.pop('ZDITHER0')
    hdulist = [primary_hdu]
    if move_1_to_0:
        starting_extension = 2
    else:
        starting_extension = 1
    for hdu in compressed_hdulist[starting_extension:]:
        if isinstance(hdu, fits.CompImageHDU):
            data_type = str(hdu.data.dtype)
            if 'int' == data_type[:3]:
                data_type = getattr(np, 'u' + data_type)
                data = np.array(hdu.data, data_type)
            else:
                data = np.array(hdu.data, hdu.data.dtype)
            header = hdu.header
            hdulist.append(fits.ImageHDU(data=data, header=header))
        else:
            hdulist.append(hdu)
    return fits.HDUList(hdulist)
```

Note above that we had to explicitly cast arrays to unsigned integers. If you do not do this, the program will crash with
```---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-85-e19442a1d873> in <module>
----> 1 astropy_unpacked_hdulist = unpack(fpacked_hdulist)

<ipython-input-84-e076fa634a7c> in unpack(compressed_hdulist)
     26             if 'int' in data_type:
     27                 data_type = '' + data_type
---> 28             hdulist.append(fits.ImageHDU(data=np.array(hdu.data), header=hdu.header))
     29         else:
     30             hdulist.append(hdu)

~/.conda/envs/fits/lib/python3.7/site-packages/astropy/io/fits/hdu/image.py in __init__(self, data, header, name, do_not_scale_image_data, uint, scale_back, ver)
   1088             data=data, header=header, name=name,
   1089             do_not_scale_image_data=do_not_scale_image_data, uint=uint,
-> 1090             scale_back=scale_back, ver=ver)
   1091 
   1092     @classmethod

~/.conda/envs/fits/lib/python3.7/site-packages/astropy/io/fits/hdu/image.py in __init__(self, data, header, do_not_scale_image_data, uint, scale_back, ignore_blank, **kwargs)
    140             # and _bscale to the appropriate BITPIX for the data, and always
    141             # sets _bzero=0 and _bscale=1.
--> 142             self.data = data
    143 
    144             # Check again for BITPIX/BSCALE/BZERO in case they changed when the

~/.conda/envs/fits/lib/python3.7/site-packages/astropy/utils/decorators.py in __set__(self, obj, val)
    753         obj_dict = obj.__dict__
    754         if self.fset:
--> 755             ret = self.fset(obj, val)
    756             if ret is not None and obj_dict.get(self._key) is ret:
    757                 # By returning the value set the setter signals that it took

~/.conda/envs/fits/lib/python3.7/site-packages/astropy/io/fits/hdu/image.py in data(self, data)
    260             # Set new values of bitpix, bzero, and bscale now, but wait to
    261             # revise original values until header is updated.
--> 262             self._bitpix = DTYPE2BITPIX[data.dtype.name]
    263             self._bscale = 1
    264             self._bzero = 0

KeyError: 'int8'
```

```python
astropy_unpacked_hdulist = unpack(fpacked_hdulist)
```

```python
astropy_unpacked_hdulist.info()
```

    Filename: (No file associated with this HDUList)
    No.    Name      Ver    Type      Cards   Dimensions   Format
      0  SCI           1 PrimaryHDU     274   (4096, 4096)   float32   
      1  CAT           1 BinTableHDU    157   306R x 37C   [D, D, D, D, K, K, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, D, K, D, D]   
      2  BPM           1 ImageHDU        10   (4096, 4096)   uint8   

```python
print(fits.FITSDiff(astropy_unpacked_hdulist, funpacked_hdulist).report())
```

    
     fitsdiff: 3.2.1
     a: <HDUList object at 0x7fa936e56818>
     b: fpacked.fits
     Maximum number of different data values to be reported: 10
     Relative tolerance: 0.0, Absolute tolerance: 0.0
    
    Primary HDU:
    
       Headers contain differences:
         Headers have different number of cards:
          a: 274
          b: 276
         Extra keyword 'BSCALE' in b: 1.0
         Extra keyword 'BZERO'  in b: 0.0
         Keyword CHECKSUM has different values:
            a> iOfhlOZgiOdgiOZg
            b> MMG4MLE2MLE2MLE2
         Keyword CHECKSUM has different comments:
            a> HDU checksum updated 2019-07-31T17:52:02
             ?                            ^ ^   ^  ^^^^
            b> HDU checksum updated 2019-08-01T00:18:54
             ?                            ^ ^  +++ ^  ^
         Keyword DATASUM  has different values:
            a> 713770948
            b> 4083115462
         Keyword DATASUM  has different comments:
            a> data unit checksum updated 2019-07-31T17:52:02
             ?                                  ^ ^   ^  ^^^^
            b> data unit checksum updated 2019-08-01T00:18:54
             ?                                  ^ ^  +++ ^  ^
    

This looks basically the same, with the only difference being the dates and checksums.

Compressing files:
================

Now we can go the other direction to make a compressed HDU. A convenience function is below:

```python
def pack(uncompressed_hdulist: fits.HDUList) -> fits.HDUList:
    if uncompressed_hdulist[0].data is None:
        primary_hdu = fits.PrimaryHDU(header=uncompressed_hdulist[0].header)
        hdulist = [primary_hdu]
    else:
        primary_hdu = fits.PrimaryHDU()
        compressed_hdu = fits.CompImageHDU(data=np.ascontiguousarray(uncompressed_hdulist[0].data),
                                           header=uncompressed_hdulist[0].header, quantize_level=64,
                                           dither_seed=2048, quantize_method=1)
        hdulist = [primary_hdu, compressed_hdu]
    
    for hdu in uncompressed_hdulist[1:]:
        if isinstance(hdu, fits.ImageHDU):
            compressed_hdu = fits.CompImageHDU(data=np.ascontiguousarray(hdu.data), header=hdu.header, quantize_level=64,
                                               dither_seed=2048, quantize_method=1)
            hdulist.append(compressed_hdu)
        else:
            hdulist.append(hdu)
    return fits.HDUList(hdulist)
```

```python
newly_compressed_hdulist = pack(uncompressed_hdulist)
newly_compressed_hdulist.writeto('astropy_fpacked.fits.fz', overwrite=True)
read_backin_astropy_compressed_hdulist = fits.open('astropy_fpacked.fits.fz')
```

```python
print(fits.FITSDiff(read_backin_astropy_compressed_hdulist, fpacked_hdulist).report())
```

    
     fitsdiff: 3.2.1
     a: astropy_fpacked.fits.fz
     b: fpacked.fits.fz
     Maximum number of different data values to be reported: 10
     Relative tolerance: 0.0, Absolute tolerance: 0.0
    
    Primary HDU:
    
       Headers contain differences:
         Headers have different number of cards:
          a: 4
          b: 8
         Extra keyword 'CHECKSUM' in b: '1hF81Z951fC51Z95'
         Extra keyword 'COMMENT' in b: "  FITS (Flexible Image Transport System) format is defined in 'Astronomy"
         Extra keyword 'DATASUM' in b: '         0'
         Keyword BITPIX   has different values:
            a> 8
            b> 16
         Keyword BITPIX   has different comments:
            a> array data type
            b> number of bits per data pixel
         Keyword EXTEND   has different comments:
            b> FITS dataset may contain extensions
         Keyword NAXIS    has different comments:
            a> number of array dimensions
            b> number of data axes
         Keyword SIMPLE   has different comments:
            a> conforms to FITS standard
             ?        -
            b> file does conform to FITS standard
             ? ++++++++++
    
    Extension HDU 1:
    
       Headers contain differences:
         Headers have different number of cards:
          a: 275
          b: 280
         Extra keyword 'BSCALE' in b: 1.0
         Extra keyword 'BZERO'  in b: 0.0
         Inconsistent duplicates of keyword 'HISTORY':
          Occurs 3 time(s) in a, 6 times in (b)
         Keyword GCOUNT   has different comments:
            a> number of groups
            b> one data group (required keyword)
         Keyword PCOUNT   has different comments:
            a> number of group parameters
            b> size of special data area
    

```python
!du -hs *.fz
```

    50M	astropy_fpacked.fits.fz
    19M	fpacked.fits.fz
    50M	fpack_straight_out.fits.fz

So I can basically bit for bit create the same fits file, but for some reason it takes 50M instead of 19M when I write it with astropy instead fpack.

```python
!du -hs --apparent-size *.fz
```

    19M	astropy_fpacked.fits.fz
    19M	fpacked.fits.fz
    19M	fpack_straight_out.fits.fz

So looking at the files it appears that there is more than a factor of 2 in padding blocks. Is it possible to remove the file padding when writing to disk? If we scp the file, the null bytes are automatically removed by the system and the file is 19M as one would expect.

```python
import os
fileSize = os.stat('astropy_fpacked.fits.fz').st_size
```

```python
fileSize
```

    19111680

```python
s = os.stat('astropy_fpacked.fits.fz')
```

```python
s.st_blksize
```

    4096

```python
s.st_blocks
```

    102016

```python
s2 = os.stat('fpacked.fits.fz')
```

```python
s2.st_blksize
```

    4096

```python
s2.st_blocks
```

    37272

So that looks like even python recognizes the padded bytes.

**Comment by cmccully** [comment]
Oddly, if you give the file instead, it appears to write the correct file size: 
```python
newly_compressed_hdulist = pack(uncompressed_hdulist)
with open('astropy_fpacked.fits.fz', 'wb') as f:
    newly_compressed_hdulist.writeto(f, overwrite=True)
read_backin_astropy_compressed_hdulist = fits.open('astropy_fpacked.fits.fz')
```

**Comment by saimn** [comment]
This was fixed by #9219, the padding was due to the incorrect dtype.
